packages/nimble-components/src/wafer-map/modules/computations.module.ts

```diff
@@ -0,0 +1,189 @@
+
+interface MapDimensions {
+    origin: {
+        x: number,
+        y: number
+    };
+    rows: number;
+    cols: number;
+}
+
+/**
+ * Computations module
+ */
+export class Computations {
+    public containerDimensions: Dimensions;
+    public dieDimensions: Dimensions = { width: 0, height: 0 };
+    public radius: number;
+    public margin: Margin;
+
+    public horizontalScale: ScaleLinear<number, number>;
+    public verticalScale: ScaleLinear<number, number>;
+
+    private readonly baseMargin: Margin = {
+        top: 20,
+        right: 20,
+        bottom: 20,
+        left: 20
+    };
+
+    private readonly dieSizeFactor = 1.5;
+    private readonly defaultAlign = 0.5;
+
+    public constructor(
+        dies: WaferMapDie[],
+        axisLocation: WaferMapQuadrant,
+        canvasDimensions: Dimensions
+    ) {
+        this.margin = this.baseMargin;
+        const gridMapDimensions = this.calculateMapDimensions(dies);
+
+        this.containerDimensions = this.calculateContainerDimensions(
+            canvasDimensions,
+            this.margin
+        );
+        this.horizontalScale = this.createHorizontalScale(
+            axisLocation,
+            gridMapDimensions,
+            this.containerDimensions.width
+        );
+        this.dieDimensions.width = this.calculateGridWidth(
+            gridMapDimensions.cols,
+            this.containerDimensions.width
+        );
+
+        this.radius = this.containerDimensions.width / 2
+            + this.dieDimensions.width * this.dieSizeFactor;
+        if (this.radius > canvasDimensions.width / 2) {
+            this.margin = this.calculateMarginWithAddition(
+                this.radius - canvasDimensions.width / 2
+            );
+            this.containerDimensions = this.calculateContainerDimensions(
+                canvasDimensions,
+                this.margin
+            );
+            this.horizontalScale = this.createHorizontalScale(
+                axisLocation,
+                gridMapDimensions,
+                this.containerDimensions.width
+            );
+            this.dieDimensions.width = this.calculateGridWidth(
+                gridMapDimensions.cols,
+                this.containerDimensions.width
+            );
+            this.radius = this.containerDimensions.width / 2
+                + this.dieDimensions.width * this.dieSizeFactor;
+        }
+
+        this.verticalScale = this.createVerticalScale(
+            axisLocation,
+            gridMapDimensions,
+            this.containerDimensions.height
+        );
+        this.dieDimensions.height = this.calculateGridHeight(
+            gridMapDimensions.rows,
+            this.containerDimensions.height
+        );
+    }
+
+    private calculateMapDimensions(dies: WaferMapDie[]): MapDimensions {
+        if (dies === undefined || dies.length === 0 || dies[0] === undefined) return { origin: { x: 0, y: 0 }, rows: 0, cols: 0 };
```

Comment:
Because the type is not optional here, we should rely on whoever creates an instance of the Computations class to check that `dies` is undefined. We don't need to duplicate that effort here.

We can simplify this to:
```suggestion
        if (dies.length === 0) return { origin: { x: 0, y: 0 }, rows: 0, cols: 0 };
```

Reply:
I understand but a problem I face is errors from ts that dies[0] may be undefined when assigning the first min and max values

Reply:
Ah, that makes sense! I didn't notice that the first index was being accessed directly. That's what I get for trying to run a TypeScript compiler in my head :P LGTM!